Improve type annotations for overloaded client methods

AlePouroullis opened the pull request:
This PR fixes the type annotations in the client overloading mechanism to preserve method return types. When users call methods like client.prompts.call(), the type system now correctly identifies the return type as PromptCallResponse instead of Any.
The implementation uses getattr/setattr to avoid type errors while maintaining identical runtime behavior.

AlePouroullis committed:
Fix type annotations in overload_client to preserve method return types

AlePouroullis committed:
Further refine types in overload.py

peadaroh commented on src/humanloop/overload.py:
```diff
@@ -184,33 +188,39 @@ def _overload_call(self: Any, sync_client: Optional[SyncClient], use_local_files
     if hasattr(client, "log") and not hasattr(client, "_log"):
-        client._log = client.log  # type: ignore[attr-defined]
+        # Store original method - using getattr/setattr to avoid type errors
+        original_log = getattr(client, "log")
```
I don't really understand why this is needed. How come it was not needed before? Could you explain why doing this avoids type errors

harry-humanloop replied:
yeah this expansion into a getattr+setattr makes it harder to read - would push against this change (as it's not the type checker pushing us to better code)

AlePouroullis replied:
Fair enough. Have addressed

peadaroh commented on src/humanloop/overload.py:
```diff
@@ -184,33 +188,39 @@ def _overload_call(self: Any, sync_client: Optional[SyncClient], use_local_files
 
-        client.log = types.MethodType(log_wrapper, client)
+        # Replace the log method
+        setattr(client, "log", types.MethodType(log_wrapper, client))
```
Same q as above for these too

harry-humanloop commented on src/humanloop/overload.py:
```diff
@@ -55,13 +59,13 @@ def _get_file_type_from_client(
+def _handle_tracing_context(kwargs: Dict[str, Any], client: T) -> Dict[str, Any]:
     """Handle tracing context for both log and call methods."""
-    trace_id = get_trace_id()
+    trace_id = get_trace_id()   
```
```suggestion
    trace_id = get_trace_id()
```

AlePouroullis committed:
refactor: simplify method overloading by using direct assignment

Replace getattr/setattr with direct attribute assignment in overload_client function. Add type ignore comments to handle type checking. This change improves code readability while maintaining the same functionality.

AlePouroullis committed:
Merge branch 'fix/specify-overload-client-return-type' of github.com:humanloop/humanloop-python into fix/specify-overload-client-return-type

AlePouroullis committed:
chore: cleaned up imports in sync_client.py
